File: src/PlinkoSimulation.java
import java.math.BigDecimal;

@SuppressWarnings({"BigDecimalMethodWithoutRoundingCalled", "SameParameterValue"})
public class PlinkoSimulation {
    private static PlinkoBoard plinkoBoard = new PlinkoBoard();
    private static PlinkoChip plinkoChip = new PlinkoChip();

    private static final int WINNER_COLUMN = 8;
    private static final int WINNER_ROW = 12;

    private static final int NUMBER_OF_TRIALS = 1000;

    public static void main(String[] args) throws Exception{
        for(DroppingLocation droppingLocation : DroppingLocation.values()){
            for(int i = 0; i < NUMBER_OF_TRIALS; i++) {
                plinkoChip.setColumn(droppingLocation);
                plinkoChip.setRow(0);
                plinkoBoard.putChipInSpot(plinkoChip);

                simulateDrop(false);

                if(isWinner()){
                    droppingLocation.incrementNumTimesDroppedIn10000();
                }
            }
        }

        printProbability();
    }

    private static void simulateDrop(boolean animate) throws Exception{
        for(int i = 0; i < PlinkoBoard.BOARD_HEIGHT - 1; i++){
            double randomNumber;

            if(plinkoChip.getColumn() == 0){
                randomNumber = 1;
            }else if(plinkoChip.getColumn() == PlinkoBoard.BOARD_WIDTH - 1){
                randomNumber = 0;
            }else{
                randomNumber = Math.random();
            }

            plinkoChip.moveChipDownOneRow();
            plinkoChip.moveColumn(randomNumber > 0.5 ? 1 : -1);
            plinkoBoard.putChipInSpot(plinkoChip);

            if(animate) {
                System.out.println(plinkoBoard);
            }
        }
    }

    private static boolean isWinner(){
        return plinkoChip.getColumn() == WINNER_COLUMN && plinkoChip.getRow() == WINNER_ROW;
    }

    private static void printProbability(){
        for(DroppingLocation droppingLocation : DroppingLocation.values()){
            BigDecimal bigDecimal = new BigDecimal(droppingLocation.getNumTimesDroppedIn10000());
            System.out.println("Probability of " + droppingLocation + " winning is: " + bigDecimal.divide(new BigDecimal(NUMBER_OF_TRIALS)));
        }
    }
}

File: python/PlinkoSimulation.py
import calendar
import time
import random

NUMBER_OF_TRIALS = 1000
NUMBER_OF_LOCATIONS = 9

BOARD_HEIGHT = 12
BOARD_WIDTH = 16

WINNER_COLUMN = 8
WINNER_ROW = 12

print("Running for " + str(NUMBER_OF_TRIALS) + " trials")

locations = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']
locationWins = [0, 0, 0, 0, 0, 0, 0, 0, 0]

chip_row = 0
chip_column = 0
random_number = 0

start = time.perf_counter()

for i in range(NUMBER_OF_LOCATIONS):
    for j in range(NUMBER_OF_TRIALS):
        chip_row = 0
        chip_column = i * 2

        for k in range(BOARD_HEIGHT):
            if chip_column == 0:
                random_number = 1
            elif chip_column == BOARD_WIDTH:
                random_number = -1
            else:
                random_number = random.random()

            chip_row += 1
            chip_column += 1 if random_number > 0.5 else -1

        if chip_row == WINNER_ROW and chip_column == WINNER_COLUMN:
            locationWins[i] += 1

end = time.perf_counter()

for i in range(NUMBER_OF_LOCATIONS):
    print("Probability of " + locations[i] + " winning is " + str(float(locationWins[i])/float(NUMBER_OF_TRIALS)))

print("Single Thread took " + str(end - start) + " seconds")
